### ISSUE TITLE
apt-key add fails when using proxies.

### ISSUE THREAD

**Issue by mmiller1** [apprenda/kismatic#976]
**Is this a BUG REPORT or FEATURE REQUEST?**:
BUG REPORT
**What happened**:
When configuring http proxies in kismatic-cluster.yaml and deploying to nodes without direct internet access apt keys / sources are not added and packages are not installed. Kismatic skips over these errors and continues deploying to the remaining hosts that do have direct internet access, ultimately failing when trying to bring up pods on the non-existent nodes. See the output from the kismatic install apply command
```
Validating==========================================================================
Reading installation plan file "kismatic-cluster.yaml"                          [OK]
Validating installation plan file                                               [OK]
Validating SSH connectivity to nodes                                            [OK]

Configuring Certificates============================================================
Found valid certificate for bizk8smaster01 etcd server                          [OK]
Found valid certificate for bizk8smaster01 API server                           [OK]
Found valid certificate for kubernetes controller manager                       [OK]
Found valid certificate for kubernetes scheduler                                [OK]
Found valid certificate for service account signing                             [OK]
Found valid certificate for bizk8smaster01 kubelet                              [OK]
Found valid certificate for kube-proxy                                          [OK]
Found valid certificate for etcd client                                         [OK]
Found valid certificate for bizk8smaster02 etcd server                          [OK]
Found valid certificate for bizk8smaster02 API server                           [OK]
Found valid certificate for bizk8smaster02 kubelet                              [OK]
Found valid certificate for bizk8smaster03 etcd server                          [OK]
Found valid certificate for bizk8smaster03 API server                           [OK]
Found valid certificate for bizk8smaster03 kubelet                              [OK]
Found valid certificate for bizhpa0s01k8s kubelet                               [OK]
Found valid certificate for bizhpb0s01k8s kubelet                               [OK]
Found valid certificate for admin client                                        [OK]
Cluster certificates can be found in the "generated" directory                  [OK]

Generating Kubeconfig File==========================================================
Generated kubeconfig file in the "generated" directory                          [OK]

Installing Cluster==================================================================
Configure Cluster Prerequisites                                                 [OK]
Gather Node Facts                                                               [OK]
Update Hosts File                                                               [SKIPPED]
Deploy Cluster Certificates                                                     [OK]
Generate Kubectl Config File                                                    [OK]
Deploy Etcd Certificates                                                        [OK]
Configure Package Repositories                                                  [ERROR]
- Task: add Docker deb key
  bizhpa0s01k8s: Failed to connect to apt.dockerproject.org at port 443: [Errno 110] Connection timed out[ERROR]
  bizhpb0s01k8s: Failed to connect to apt.dockerproject.org at port 443: [Errno 110] Connection timed out[ERROR]
Configure Package Repositories                                                  [OK]
Install Docker                                                                  [OK]
Start Kubernetes Etcd Cluster                                                   [OK]
Start Network Etcd Cluster                                                      [OK]
Start Kubernetes Kubelet                                                        [OK]
Start Kubernetes API Server                                                     [OK]
Start Kubernetes Scheduler                                                      [OK]
Start Kubernetes Controller Manager                                             [OK]
Validate Kubernetes Control Plane is Running                                    [OK]
Start Kubernetes Proxy                                                          [OK]
Label Kubernetes Nodes                                                          [SKIPPED]
Start Calico Network Components                                                 [OK]
Validate Calico Network Components                                              [OK]
  bizk8smaster01: Timed out waiting for policy controller pods to be in the ready state.[ERROR]
error installing: error running playbook: error running ansible: exit status 2
```
In this case, my control plane is on virtual machines with direct internet access, and the core kubernetes components are installed as expected, though the workers are on physical machines with more restrictive networking and all package deployment fails, despite passing preflight checks.
**What you expected to happen**:
1.) I expected kismatic to stop attempts to deploy components when no workers had been created.
2.) I expected apt-key and package installations to work properly when setting http_proxy & https_proxy values.
**How to reproduce it (as minimally and precisely as possible)**:
Deploy workers in environments that require proxies for external communication & master nodes in environments that do not.
Set http_proxy & https_proxy in kismatic-cluster.yaml
run kismatic install apply
**Anything else we need to know?**:
The proxy settings for apt appear to work on the boxes that require it. I can run all of the apt commands with sudo as the kismatic user with success. It is just when being run from the kismatic deploy box that things fail.
**Environment**:
- Kismatic version (use `kismatic version`): 1.6.3
- Plan file Gist (remove sensitive information if any): https://gist.github.com/msm93v2/4beace0cba8ddbff9c5cb7b28076f6f1
- Cloud provider or hardware configuration: bare-metal, control plan on vms, physical device workers.
- OS (e.g. from /etc/os-release): Ubuntu 16.04
- Kernel (e.g. `uname -a`): 4.4.X

**Comment by dkoshkin** [comment]
Hi @msm93v2 thank you for opening the issue.
The proxy options in the plan file are global and set the environmental variables on all the nodes.

You can try setting the variables manually on the nodes that need it:
```
  HTTPS_PROXY: "{{ https_proxy }}"
  https_proxy: "{{ https_proxy }}"
  HTTP_PROXY: "{{ http_proxy }}"
  http_proxy: "{{ http_proxy }}"
  NO_PROXY: "{{ no_proxy }}"
  no_proxy: "{{ no_proxy }}" 
```

The other comment about the installation proceeding after a task failed is more concerning, I will take a look at that.

**Comment by msm93v2** [comment]
Setting these variables globally does not change the behavior of `kismatic
install apply`. perhaps the issue is not in the proxy settings itself, but
something else. Are there debug logs that my be more helpful in finding the
issue?

**Comment by dkoshkin** [comment]
From your description it would be the proxy settings that are being used on the nodes where the proxy should not be used. By "global" I meant those options will be used on all the nodes during the installation, it will not persist the variables after the installation.

Are you able to configure your network that all the nodes are able to either all use the proxy or all not use the proxy?

There should be logs in the `generated\` directory for Ansible and you can find the `add Docker deb key` task.

**Comment by msm93v2** [comment]
The nodes that require the proxy are failing to talk out, the nodes that
don't *need* the proxy can talk through out just fine with the proxy set.
As I mentioned, these commands run fine in both login, and non-login shells
on all nodes. I have worked around the issue using the offline installation
method though, so if no one else has issues with it, I'm perfectly happy
chalking it up to an issue with my environment.

**Comment by dkoshkin** [comment]
Closed in #1173
